pkg/storage/common/common.go

```diff
@@ -529,7 +531,104 @@ func IsBlobReferenced(imgStore storageTypes.ImageStore, repo string,
 		return false, err
 	}
 
-	return isBlobReferencedInImageIndex(imgStore, repo, digest, index, log)
+	return IsBlobReferencedInImageIndex(imgStore, repo, digest, index, log)
+}
+
+/* Garbage Collection */
+
+func AddImageManifestBlobsToReferences(imgStore storageTypes.ImageStore,
+	repo string, mdigest godigest.Digest, refBlobs map[string]bool, log zlog.Logger,
+) error {
+	manifestContent, err := GetImageManifest(imgStore, repo, mdigest, log)
+	if err != nil {
+		log.Error().Err(err).Str("repository", repo).Str("digest", mdigest.String()).
+			Msg("gc: failed to read manifest image")
+
+		return err
+	}
+
+	refBlobs[mdigest.String()] = true
+	refBlobs[manifestContent.Config.Digest.String()] = true
+
+	// if there is a Subject, it may not exist yet and that is ok
+	if manifestContent.Subject != nil {
+		_, err := GetImageManifest(imgStore, repo, manifestContent.Subject.Digest, log)
```

Comment:
I think the correct questions to ask are:
- is this manifest digest tagged?
- is this manifest digest referring to a tagged manifest?
- is this manifest digest referring to a manifest with is referring to a tagged manifest?
- and so on
Not sure if this is the right place to add that logic or not.

Reply:
I think one of the issues is, if you use only references and created times, like umoci does, you basically assume someone deletes unwanted manifests along with the tags in the DELETE API call, and in that case it is easy, you just search for all orphan (unreferenced) blobs and delete them.

But since there is the case you can delete a tag, but not the manifest, the umoci behavior is not enough. We want the user to delete the tag, and GC delete the manifests behind the scenes, if it is truly unused. So we need a smarter way to do this cleanup.

Reply:
nice catches.
ok, let me think about it.